# Traffic Sign Classification - LeNet Deep Network

### by ReDay Zarra

This project utilizes the LeNet deep network architecture to classify 42 different types of traffic signs. LeNet refers to a convolutional neural network that can be used for computer vision and classification models. This project showcases a step-by-step implementation of the model as well as in-depth notes to customize the model further for higher accuracy.

## Importing the necessary libraries

In [36]:
import pandas as pd
import numpy as np

> Classic libraries that will help us read and analyze data

In [37]:
import matplotlib.pyplot as plt
import seaborn as sns

> Libraries used for plotting and data visualization

In [38]:
import pickle

> Using pickle package to open our data, not much use after that

## Importing the dataset

Importing the data into three different sections: training, validation, and
testing. Training data is used to train the network, testing data is used to test
the trained network with data that it has never seen to see how it would perform in
the real world. The validation dataset ensures that we avoid overfitting by showing
the network validation data every epoch (or run) in a process called 
cross-validation to ensure that the network is not focusing on the details of the
training data. 

In [39]:
with open("./traffic-signs-data/train.p", mode = 'rb') as training_data:
    train = pickle.load(training_data)

In [40]:
with open("./traffic-signs-data/valid.p", mode = 'rb') as validation_data:
    valid = pickle.load(validation_data)

In [41]:
with open("./traffic-signs-data/test.p", mode = 'rb') as testing_data:
    test = pickle.load(testing_data)

### Splitting the dataset into training and testing sets

Splitting the data into our individual training, validation, and testing set variables.

In [42]:
X_train, y_train = train['features'], train['labels']

> Assigning the features of the training set as X_train and the dependent variable (labels) as y_train

In [43]:
X_validation, y_validation = valid['features'], valid['labels']

> Assigning the features of the validation set as X_validation and the dependent variable (labels) as y_validation

In [44]:
X_test, y_test = test['features'], test['labels']

> Assigning the features of the testing set as X_test and the dependent variable (labels) as y_test

### Checking the dimensions of the datasets

In [45]:
X_train.shape

(34799, 32, 32, 3)

> Gives us an output of a four element tuple. The first number is the quantity of images, the second is the width of image in pixels, the third is the height of the image, and the last number the depth - in this case the 3 tells us that the images are colored since they are being multiplied for both Red, Green, and Blue

In [46]:
y_train.shape

(34799,)

> Gives us a tuple of one element, which is a label for each image in the training set

In [47]:
X_validation.shape

(4410, 32, 32, 3)

> A smaller subset only containing 4410 pictures, which are 32 x 32, and have a depth of 3 meaning they are colored with RGB

In [48]:
X_test.shape

(12630, 32, 32, 3)

In [49]:
y_test.shape

(12630,)